- [1/6](#16)
  - [科研-氮肥文章](#科研-氮肥文章)
    - [氮肥讨论](#氮肥讨论)
      - [堆叠图](#堆叠图)
    - [文章修改](#文章修改)
      - [安装graphpad](#安装graphpad)
      - [introduction文献](#introduction文献)
  - [1/7](#17)
    - [氮肥](#氮肥)
      - [图](#图)
      - [introduction](#introduction)
- [1/8](#18)
  - [氮肥文章](#氮肥文章)
    - [introduction](#introduction-1)
    - [p图](#p图)
      - [热图已完成](#热图已完成)
      - [箱型图](#箱型图)
- [1/9](#19)
  - [氮肥文章](#氮肥文章-1)
    - [改图](#改图)
      - [正文图](#正文图)
      - [附近文件图](#附近文件图)
- [1/10](#110)
  - [G4](#g4)
    - [LTR_finder](#ltr_finder)
    - [tmux](#tmux)
      - [安装及教程](#安装及教程)
      - [Tmux查看会话历史输出信息](#tmux查看会话历史输出信息)
    - [LTR_finder](#ltr_finder-1)
- [1/14](#114)
  - [氮肥文章](#氮肥文章-2)
    - [完成图注](#完成图注)
    - [检查文章](#检查文章)
  - [G4](#g4-1)
    - [文献阅读](#文献阅读)
      - [Gene-LTR-retrotransposons in plants: Engines of evolution](#gene-ltr-retrotransposons-in-plants-engines-of-evolution)
      - [重要信息](#重要信息)
      - [总结研究思路](#总结研究思路)
- [1/15](#115)
  - [G4](#g4-2)
    - [que 如何构建进化树](#que-如何构建进化树)
      - [step1](#step1)
    - [找失活的转座子想法](#找失活的转座子想法)
      - [step1 建树](#step1-建树)
      - [step2 分类群 测K_ST](#step2-分类群-测k_st)
      - [step3](#step3)
- [1/17](#117)
  - [BLAST](#blast)
  - [处理.ltr文件](#处理ltr文件)
- [1/19](#119)
  - [G4](#g4-3)
    - [ltr](#ltr)
      - [pipeline-TE.py完善](#pipeline-tepy完善)
# 1/6
## 科研-氮肥文章
### 氮肥讨论
#### 堆叠图
讨论无植物和有植物的微生物排序问题
找与氮肥有关的菌的变化

先说堆叠图
再说bacilli
introduction也要加重点说bacilli


一热图加一个表
另外两个热图＋附件文件

### 文章修改
#### 安装graphpad

#### introduction文献
1 介绍bacillus
https://www.frontiersin.org/articles/10.3389/fmicb.2019.01424/full

 2 微生物搭便车现象--主要是运输链霉菌孢子
https://www.nature.com/articles/s41396-021-00952-8

3 微生物搭便车现象
https://www.cell.com/trends/microbiology/fulltext/S0966-842X(20)30267-5?rss=yes
 

 ## 1/7
 ### 氮肥
 #### 图
 两两比较 箱型图
 文件名：2to2.pzfx
 #### introduction

 paragraph1 bacillus
 
Bacillus is one of the important bacteria that hitchhiking phenomenon occurs in soil, for example, Streptomyces spores are transported to plant tissues by Streptomyces spores attach to Bacillus.

 paragraph3 

 There is clear evidence that plants shape microbiome structures, most probably by root exudates, and also that bacteria have developed various adaptations to thrive in the rhizospheric niche.  Besides, plants block a majority of microbes at the front line by a non-host resistance strategy, this natural immunity affects the growth of microbial populations in the soil. These studies suggest that plant research influences microbial communities, but no relevant studies have revealed the relationship between plants and hitchhiking.

 **植物影响微生物群落，字数不够，保证每段字数差不太多，举例说明，比如说某种菌是如何被植物影响的**

 paragraph4 
 
 研究说明：改成植物和氮肥协同作用，删掉真菌的部分

 We proposed that under the synergistic effect of nitrogen application concentration and plant, rhizosphere microbial migration and hitchhiking activities were affected, leading to changes in rhizosphere bacterial communities, and thus affecting plant growth.We present findings from a long-term longitudinal field study, in which three different nitrogen fertilizer concentrations were applied to different groups, and 0.45-μm filter membranes were utilized to prevent microbe movement through the soil. We used a molecular barcoding approach to assess bacterial diversity patterns using DNA directly extracted from bulk soil as the template to amplify the 16S rRNA gene region by PCR.Then, high-throughput sequencing of the obtained amplicons was performed to illustrate the changes in the bulk soil and rhizosphere microbial communities under different nitrogen concentrations and with or without spinach planting . Furthermore, the mechanism responsible for the effects of filter membrane applications near rhizosphere soil exposed to various nitrogen concentrations and with or without spinach planting on plant growth was investigated. We assessed the current co-occurrence network and studied the activation of the symbiotic mode under various nitrogen concentrations and with or without spinach cultivation, and we found that nitrogen concentration and with or without spinach cultivation changes influenced both motile and their associated non-motile bacteria.


**最后检查一下逻辑**


# 1/8
## 氮肥文章
### introduction
完善例子
Maize plants release a blend of metabolites, including BXs , from the roots and thereby influence the composition of the root-associated microbiota.

https://www.nature.com/articles/s41467-018-05122-7\

The composition of root exudates varies from plant to plant and affects the relative abundance of microorganisms in the vicinity of the root, plants not only can provide nutrients for microorganisms, but some plant species also contain unique antimicrobial metabolites in their exudates.  

https://academic.oup.com/femsec/article/68/1/1/572572
修改完成

### p图
#### 热图已完成
#### 箱型图
软件有问题 所以

**reason1**

Previous studies have suggested that root length and root surface area were increased under intermediate N levels and that root growth was reduced under both higher and lower fertilization levels. However, high N availability and its concomitant affect root and shoot biomass production

https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0171321

Nitrogen is known to be functional in the construction of amino acids and chlorophyll, which can influence plant growth and development by affecting photosynthesis and the uptake of minerals [25,26]. 
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7024162/#!po=88.7500

Previous studies have suggested that root length and root surface area were increased under intermediate N levels and that root growth was reduced under both higher and lower fertilization levels, so that the plant morphology is affected. On the other hand, Nitrogen is be functional in the construction of amino acids and chlorophyll, which can influence plant growth and development by affecting photosynthesis and the uptake of minerals.  


# 1/9
## 氮肥文章
### 改图
#### 正文图
#### 附近文件图
**差一个12图示意图 需要具体说明膜**
**相互作用图6图**



# 1/10
## G4
### LTR_finder
安装 需要配置环境变量
1 添加环境变量
vim ~/.bashrc 
export PATH="/usr/local/nginx/sbin/:$PATH"
退出
source ~/.bashrc
### tmux
#### 安装及教程
https://www.ruanyifeng.com/blog/2019/10/tmux.html
#### Tmux查看会话历史输出信息
然后先按快捷键ctrl+b ，然后再按下[，即进入历史输出信息查看模式，可通过键盘上的上下左右键来滚动历史输出信息。
如果要退出查看模式，按下q即可。

### LTR_finder

关于LTR和LTR_retriever软件的介绍
https://www.jianshu.com/p/8fa4ef437870



# 1/14
## 氮肥文章
### 完成图注
### 检查文章
**示意图刘老师还没看（意思是一整天都没见到刘老师）**

## G4
### 文献阅读
#### Gene-LTR-retrotransposons in plants: Engines of evolution
http://dx.doi.org/10.1016/j.gene.2017.04.051

#### 重要信息
1 拟南芥中的转座子占拟南芥基因组的 14%以上

2 当转座子以相反方向插入基因内部/靠近基因产生反义转录物/通过将 TEs甲基化延伸到侧翼区域--基因沉默

3 LTR是造成 C 值悖论的主要原因

4 转座子的插入可能会导致--基因功能的改变
插入外显子/内含子；表观遗传沉默；影响转录（超过其正常的终止序列）

#### 总结研究思路
![avatar](/七崽工作进展/Figure/思路.png)


# 1/15
## G4
### que 如何构建进化树
**RAxML构建进化树** 
软件--RAxML
#### step1
文件处理
先将ltr_finder的ltr文件转为bed文件
`grep -v "Sequence" chr1.ltr |grep "Location" -B 1 |sed 's/--//g' |sed '/^$/d' |awk '{if(substr($1,1,1)=="[") {print $2} else {print $3,$5,$8}}' |sed 's/Strand://g' |tr "\n" "\t" |sed 's/+/+\n/g' |sed 's/-/-\n/g' |awk '{if(NR==1) {print "\t"$0} else {print $0}}' |sed 's/\t//1' |tr " " "\t" >chr1.bed`

该软件是对fasta文件进行处理的
需要把转座子的bed文件转成fasta文件

该bed文件不包括序列信息 所以需要利用bedtools提取对应的序列信息 并存为fasta文件
bedtools-getfasta
```bedtools getfasta -fi test.fa -bed test.bed -s -fo result.fa ```
-s 考虑链的正负方向
https://www.jianshu.com/p/5c053064fab9


### 找失活的转座子想法
#### step1 建树
以LTR各种家族为分类标准 判断不同家族的分化情况
![avatar](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/gbe/5/5/10.1093_gbe_evt064/4/m_evt064f3p.jpeg?Expires=1645184951&Signature=HcBT5ZgWBuNdOvtfSOfiqSM7cOZU9uDmXHT~v64r4B-bZzB32i7MoqAUnou9BlUAALLngjFeW45Xpz0iyq7SpQ6plDV9qICMuo-iEsvVZJ38sdmhoOmsXLN3szwInBB8PMmLSVWJJWRbXDylsunSCWGk5791bXrcy2wiRwMhkNvLo0PyIuKJ2Ek69RtqgDxdikctTilEYFFY9pIohLHIvOH4fOmSzO1s4x~b1qG3-5flifeWU72RVCh8GXgjeyp3UGgtXcDxaJ8aW06-uPq3keaTN66mAufxqPWvO~~lbjmBulQ4eLdgb0aY-ra5Gzts4dDMODxq8xKEuqWexd9COQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA)

#### step2 分类群 测K_ST
kst表示以划分的类群1和2之间的遗传分化
中间的kst值表示 物种间共享相似的TE拷贝
![avatar](../Figure/table.png)

#### step3 
具有混合物种进化枝（树 IV）和低K ST的ML树
表明物种分化期间的该LTR家族的静止


# 1/17
## BLAST
+ 用于序列分析，同源的数据库搜索软件
+ BLAST软件包可以 用于
  + BLASTP 蛋白质序列 vs 蛋白质库
  + **BLASTN** DNA序列 vs DNA库
  + BLASTX DNA序列 vs 蛋白质库
+ 和FASTA比较 BLAST比较快

## 处理.ltr文件
+ 将ltr转为bed文件
+ 源代码
```python
from cgi import test
from imghdr import tests
from tkinter import W
import re

def ltr_bed(ltrfile,bedfile):
    ltrfile=open(ltrfile,'r')
    bedfile=open(bedfile,'w')
    i=0
    a=0
    q=0
    chrname=[]
    startnumber=[]
    endnumber=[]
    style=[]
    pattern0 ='(?<=chr).[0-9]*'
    pattern1 = '(?<=Location : ).[0-9]*'
    pattern2 = '(?<= - ).[0-9]*'
    pattern3 = '(?<=Strand:).[0-9]*'
    for line in ltrfile:
        # 把line分割为每个转座子信息为一个字符串 line.split("--")
        # re.search()扫描整个字符串并返回第一个成功的匹配。
        # 以空格为分隔符 提取chrx
        if 'chr'in line:
            chrname.append(re.findall(pattern0, line))
            test0 = "".join(chrname[q])
            bedfile.write('chr'+test0+'-'+str(a+1)+'  ')
            q=q+1

        m='Location'
        if m in line:
            # 正则表达式处理
            startnumber.append(re.findall(pattern1, line))
            test1 = "".join(startnumber[a])
            bedfile.write(test1+'  ')

            endnumber.append(re.findall(pattern2, line)[0])
            test2 = "".join(endnumber[a])
            bedfile.write(test2+'  ')

            style.append(re.findall(pattern3, line))
            test3 = "".join(style[a])
            bedfile.write(test3+'\n')
            a=a+1
        i=i+1     
    ltrfile.close
    bedfile.close

if __name__ == '__main__':
    ltrfile="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/test.ltr"
    bedfile="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/1x.bed"
    ltr_bed(ltrfile,bedfile)          
```

+ 测试输出结果
染色体-转座子号 起始位点 结束位点 正/负链
chr1-1  179795  184613  -
chr1-2  255305  264911  -
chr1-3  435375  443647  +
chr1-4  530597  536514  +
chr1-5  530597  536882  +
chr1-6  530597  537224  +
chr1-7  530597  537848  -

+ 写代码遇到的问题 
  + 对```for line in file```不够熟悉
  + 该操作是对每行进行读取，一行一行的读，把整个文件扫一遍，理解有误
  + 对正则表达式不够熟练，``` '(?<= - ).[0-9]*'```





# 1/19
## G4
### ltr
#### pipeline-TE.py完善
```python
from cgi import test
from imghdr import tests
from platform import system
from tkinter import W
import re
import os


# 处理ltr文件  ltr-bed
def ltr_bed(ltrfile,bedfile):
    ltrfile=open(ltrfile,'r')
    bedfile=open(bedfile,'w')
    i=0
    a=0
    q=0
    # chrid=[]
    chrname=[]
    startnumber=[]
    endnumber=[]
    style=[]
    pattern0 ='(?<=chr).[0-9]*'
    pattern1 = '(?<=Location : ).[0-9]*'
    pattern2 = '(?<= - ).[0-9]*'
    pattern3 = '(?<=Strand:).[0-9]*'
    for line in ltrfile:
        # 把line分割为每个转座子信息为一个字符串 line.split("--")
        # re.search()扫描整个字符串并返回第一个成功的匹配。
        # 以空格为分隔符 提取chrx
        if 'chr' in line:

            chrname.append(re.findall(pattern0, line))

            test0 = "".join(chrname[q])
            bedfile.write('chr'+test0+'\t')
            q=q+1

        m='Location'
        if m in line:
        # if (line.split("--")[i].find("str") != -1): 
            # 正则表达式处理   
            
            startnumber.append(re.findall(pattern1, line))
            test1 = "".join(startnumber[a])
            bedfile.write(test1+'\t')

            endnumber.append(re.findall(pattern2, line)[0])
            test2 = "".join(endnumber[a])
            bedfile.write(test2+'\t')

            style.append(re.findall(pattern3, line))
            test3 = "".join(style[a])
            bedfile.write(test3+'\n')
            a=a+1
            # test = "".join(chrname[q-1])
            # bedfile.write('chr'+test+'-'+str(a+1) +'\n')
        i=i+1     
    ltrfile.close
    bedfile.close

def modifyfasta(fastafile,fastafile1):
    chrname=[]
    chrnumber1=[]
    chrnumber2=[]
    q=0

    fastafile = open(fastafile,'r')   
    fastafile1 = open(fastafile1,'w') 
    i=1
   
    find0='(?<=:).[0-9]*'
    find1='(?<=-).[0-9]*'
    pattern1 ='(?<=chr).[0-9]*'

    for line in fastafile:
        if line.startswith('>'):
            # step1 提取chr后的染色体数
            chrname.append(re.findall(pattern1, line))
            test1 = "".join(chrname[q])
            fastafile1.write('>'+'chr'+test1+'-'+str(i)+':')

            chrnumber1.append(re.findall(find0, line))
            chrnumber2.append(re.findall(find1, line))
            test3="".join(chrnumber1[q])
            test4="".join(chrnumber2[q])
            fastafile1.write(test3+'-'+test4+'()'+'\n')
            
            i=i+1
            q=q+1
            continue
       
        fastafile1.write(line)
    fastafile.close()
    fastafile1.close()




if __name__ == '__main__':
    chr_fastafile="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/chr1.fa"
    ltrfile="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/test.ltr"
    bedfile="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/1x.bed"
    fastafile="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/resultx.fa"
    fastafile1="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/resultx-modify.fa"
    ltr_identify="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/An-1/ltr_identify.out"
    # step1 将ltr文件转为bed文件 
    ltr_bed(ltrfile,bedfile) 
    # step2 将bed文件转为fasta文件 - getfasta
    os.system("bedtools getfasta -fi "+chr_fastafile+" -bed "+bedfile+" -s -fo "+fastafile)
    # step3 将fasta文件进行一些处理 使得最终ltr的命名以chr染色体号-第几个ltr转座子
    modifyfasta(fastafile,fastafile1)
    # step4 对ltr进行balst操作
    os.system("blastn -query "+fastafile1+" -db /home/ubuntu/data/softwares/tedb -outfmt 6 > "+ltr_identify)
    # step4 处理blastn后的文件
    # que：选择哪一个类别作为该ltr的最终类别？
    

```

+ blastn结果
```xml
chr1-3:435375-443647()	AT3TE36695|-|8752366|8752789|SIMPLEGUY1|DNA/Harbinger|424	82.963	135	10	5	2644	2765	222	88	1.36e-22	110
chr1-3:435375-443647()	AT1TE96850|+|29611562|29611719|TAG2|DNA/HAT|158	85.526	152	20	2	2384	2535	1	150	4.78e-37	158
chr1-3:435375-443647()	AT1TE53880|-|16351341|16351942|SIMPLEGUY1|DNA/Harbinger|602	75.749	367	57	18	2564	2923	423	82	1.72e-36	156
chr1-3:435375-443647()	AT2TE10335|+|2233114|2233258|TAG2|DNA/HAT|145	86.275	153	7	4	2977	3129	1	139	6.18e-36	154
chr1-3:435375-443647()	AT3TE41760|-|10031935|10031978|SIMPLEGUY1|DNA/Harbinger|44	100.000	44	0	0	2921	2964	44	1	2.96e-14	82.4
chr1-4:530597-536514()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	99.475	381	2	0	5120	5500	1	381	0.0	693
chr1-4:530597-536514()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	100.000	34	0	0	5583	5616	383	416	7.66e-09	63.9
chr1-5:530597-536882()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	99.475	381	2	0	5120	5500	1	381	0.0	693
chr1-5:530597-536882()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	100.000	34	0	0	5583	5616	383	416	8.14e-09	63.9
chr1-6:530597-537224()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	99.475	381	2	0	5120	5500	1	381	0.0	693
chr1-6:530597-537224()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	100.000	34	0	0	5583	5616	383	416	8.58e-09	63.9
chr1-7:530597-537848()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	99.475	381	2	0	5120	5500	1	381	0.0	693
chr1-7:530597-537848()	AT1TE01710|+|529161|529576|HELITRON4|RC/Helitron|416	100.000	34	0	0	5583	5616	383	416	9.39e-09	63.9
```
+ pipeline遗留问题：选择哪一个类别作为该ltr的最终类别？
  + 对identify.out文件的处理
  + 是选择分数最高的（可能性最高）的作为最终类别还是可能的都计算1遍