# 4/1
## g4
### 进化树抽样
+ 每个xml生成random.xml，然后利用sort对其进行排序
  + -k 2.6：第二个元素的第六个字符开始 
  + -t：分隔符，tab比较特殊 如果不转义会出现multi-character tab ‘\t’
    + 解决方案1：-t$'\t' 
    + 解决方案2：-t'<ctrl>v<tab>'
+ 抽样方法：0-50的随机数生成
``` python 
def randomSample(ltr_family_getfile,random_xml_file):
    #count total
    # number=os.system("awk 'END {print NR}' "+ltr_family_getfile)
    # random
    ran_number=random.randint(0,50)

    os.system("shuf -n"+str(ran_number)+" "+ltr_family_getfile+">"+random_xml_file)

def sortSample(random_xml_file,random_file):
    os.system("bash /home/ubuntu/Arabidopsis/Scripts/sort.sh "+random_xml_file+" "+random_file)
```
``` bash
# $1 xml_path
# /home/ubuntu/Arabidopsis/Arabidopsis_sequence/159/chr5.random-ltr.xml
# $2 xml_random_path
# /home/ubuntu/Arabidopsis/Arabidopsis_sequence/159/chr5.random.ltr.xml
sort -n -u -k 2.6 -t$'\t' $1 > $2
# sort -n -u -k 2.6 -t'<ctrl>v<tab>' $1 > $2
```
+ tmux kill session
  + 使用会话编号:tmux kill-session -t 0
  + 使用会话名称:tmux kill-session -t session7

### sequence_2完成
+ 跑代码
  + blast
  + random sample


# 4/2
## 生信作业
+ blast
+ 2-4次作业完善
```python

import os


def dealBlast(savefile,blastfile):
    #读取文件
    infile-open(savefile,'r')
    outfile=open(blastfile,'w')
    test=''
    for line in infile:
        inline=line.split('\t')
        if(str(inline[0])==''):
            continue
        if(test==str(inline[0])):
            continue
        outfile.write(line)
        test=str(inline[0])



if __name__ == '__main__':
    # step1 建立数据库
    dbfile="/home/ubuntu/Arabidopsis/1.fasta"
    os.system("makeblastdb -in "+dbfile+" -dbtype prot -parse_seqids -hash_index -out uniprot")

    # setep2 做blast
    g = os.walk(r"e:\test")  
    names="/home/ubuntu/Arabidopsis/Arabidopsis_sequence/test/"
    i=2

    # while i<13:
    while i<13:
        # name:filename
        infile=names+str(i)+'.fasta'
        savefile=names+str(i)+".blast"
        blastfile=names+str(i)+".mofidy.blast"
        os.system("blastp -query "+infile+" -db uniprot -out "+savefile)
        i=i+1
        # 处理balst文件 得到evalue最低的配对序列
        dealBlast(savefile,blastfile)
  ```
+ 待完成：蛋白翻译，6个读码框（正义链和反义链 2*3）
  + 找到最长的翻译结果
  + 参考网站https://web.expasy.org/translate/
## ltr——classification
+ list文件填写

# 4/3
## g4
+ 完成ltr+classification两个pipeline的编写，随机抽样创建fasta文件
+ 问题:建树效率十分低，无法在短时间内完成建树工作
```python classification
import os

import re

def getLtrList(paths,path,familys):
    # 创建文件存储转座子的信息
    familyfile0=open(paths+'LTR-Gypsy.txt','w')
    familyfile1=open(paths+'LTR-Copia.txt','w')
    familyfile2=open(paths+'DNA-En-Spm.txt','w')
    chr_number=1
    while chr_number<=5:
        xmlfile=open(path+'/chr'+str(chr_number)+'.random.ltr.xml','r')
        for line in xmlfile:
            test=line.split('\t')
            test0=str(test[2]).split('\n')
            if(str(test0[0])==str(familys[0])):
                familyfile0.write(str(test[1])+'\n')

            if(str(test0[0])==str(familys[1])):
                familyfile1.write(str(test[1])+'\n') 

            if(str(test0[0])==str(familys[2])):
                familyfile2.write(str(test[1])+'\n')

        chr_number+=1
    familyfile0.close()
    familyfile1.close()
    familyfile2.close()
    xmlfile.close()


def getLtrFastaInfo(paths,path):
    fastafile0=open(paths+'LTR-Gypsy.fasta','w')
    fastafile1=open(paths+'LTR-Copia.fasta','w')
    fastafile2=open(paths+'DNA-En-Spm.fasta','w')

    familyfile0=open(paths+'LTR-Gypsy.txt','r')
    familyfile1=open(paths+'LTR-Copia.txt','r')
    familyfile2=open(paths+'DNA-En-Spm.txt','r')
    # familyfile0=open(familyfile0,'r')
    # familyfile1=open(familyfile1,'r')
    # familyfile2=open(familyfile2,'r')

    chr_number=1

    while chr_number<=5:
        fastafile=open(path+'/chr'+str(chr_number)+'.ltr.fasta','r')
        x=str(path).split('/')
        sequence_number=str(x[5])
        
        for line in fastafile:
            if line.startswith('>'):
                a=line.split('>')
                continue

            # for line0 in familyfile0:
            #     if(str(line0)==str(a[1])):
            #         #写入文件 
            #         fastafile0.write('>'+sequence_number+'-'+str(a[0])+str(line0))

            # for line1 in familyfile1:
            #     if(str(line1)==str(a[1])):
            #         fastafile1.write('>'+sequence_number+'-'+str(a[0])+str(line1))
            # for line2 in familyfile2:
            #     if(str(line2)==str(a[1])):
            #         fastafile2.write('>'+sequence_number+'-'+str(a[0])+str(line2))
        chr_number=chr_number+1
    fastafile.close()
    familyfile0.close()
    familyfile1.close()
    familyfile2.close()

def getFastaInfor(path,paths,familys):

    # 创建文件存储转座子的信息
    familyfile0=open(paths+'LTR-Gypsy.txt','a+')
    familyfile1=open(paths+'LTR-Copia.txt','a+')
    familyfile2=open(paths+'DNA-En-Spm.txt','a+')

    fastafile0=paths+'LTR-Gypsy.fasta'
    fastafile1=paths+'LTR-Copia.fasta'
    fastafile2=paths+'DNA-En-Spm.fasta'



    chr_number=1
    while chr_number<=5:
        ltr_number=''
        ltr_number0=''
        xmlfile=path+'/chr'+str(chr_number)+'.random.ltr.xml'
        fastafile=path+'/chr'+str(chr_number)+'.ltr.fasta'
        with open(xmlfile,'r') as f1, open(fastafile,'r') as f2, open(fastafile0,'a+') as wp1,open(fastafile1,'a+') as wp2,open(fastafile2,'a+') as wp3:
            for l2 in f2:
                if(l2.startswith(">")):


                    if(ltr_number==ltr_number0):
                        l1 = f1.readline()
                        if(l1==''):
                            break
                        
                        ltr_number0=str(l1.split("\t")[1])
                        fa_number0=str(l1.split("\t")[2])
                        fa_number=str(fa_number0.split('\n')[0])
                        group_number=str(l1.split("\t")[0])

                    test=str(l2.split(">")[1])
                    ltr_number=str(test.split('\n')[0])
                        

                    if(ltr_number==ltr_number0):
                        if(fa_number==str(familys[0])):
                            wp1.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile0.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                            

                        if(fa_number==str(familys[1])):
                            wp2.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile1.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')

                        if(fa_number==str(familys[2])):
                            wp3.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile2.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                        continue
                    else:
                        continue

                if(ltr_number==ltr_number0):
                    if(fa_number==str(familys[0])):
                        wp1.write(l2)
                    if(fa_number==str(familys[1])):
                        wp2.write(l2)                    
                    if(fa_number==str(familys[2])):
                        wp3.write(l2)
        chr_number=chr_number+1

def getReFastaInfor(paths,familys):
    # 创建文件存储转座子的信息
    familyfile0=open(paths+'LTR-Gypsy.txt','w')
    familyfile1=open(paths+'LTR-Copia.txt','w')
    familyfile2=open(paths+'DNA-En-Spm.txt','w')

    fastafile0=paths+'LTR-Gypsy.fasta'
    fastafile1=paths+'LTR-Copia.fasta'
    fastafile2=paths+'DNA-En-Spm.fasta'

    ltr_number=''
    ltr_number0=''

    xmlfile=paths+'rubella.random.ltr.xml'
    fastafile=paths+'.ltr.fasta'
    with open(xmlfile,'r') as f1, open(fastafile,'r') as f2, open(fastafile0,'w') as wp1,open(fastafile1,'w') as wp2,open(fastafile2,'w') as wp3:
        for l2 in f2:
                if(l2.startswith(">")):


                    if(ltr_number==ltr_number0):
                        l1 = f1.readline()
                        if(l1==''):
                            break
                        
                        ltr_number0=str(l1.split("\t")[1])
                        fa_number0=str(l1.split("\t")[2])
                        fa_number=str(fa_number0.split('\n')[0])
                        group_number=str(l1.split("\t")[0])

                    test=str(l2.split(">")[1])
                    ltr_number=str(test.split('\n')[0])
                        

                    if(ltr_number==ltr_number0):
                        if(fa_number==str(familys[0])):
                            wp1.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile0.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')
                            

                        if(fa_number==str(familys[1])):
                            wp2.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile1.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')

                        if(fa_number==str(familys[2])):
                            wp3.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile2.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')
                        continue
                    else:
                        continue

                if(ltr_number==ltr_number0):
                    if(fa_number==str(familys[0])):
                        wp1.write(l2)
                    if(fa_number==str(familys[1])):
                        wp2.write(l2)                    
                    if(fa_number==str(familys[2])):
                        wp3.write(l2)






def writeReference(referencePath,paths):
    # familyfile0=open(paths+'LTR-Gypsy.txt','a+')
    # familyfile1=open(paths+'LTR-Copia.txt','a+')
    # familyfile2=open(paths+'DNA-En-Spm.txt','a+')
    # fastafile0=open(paths+'LTR-Gypsy.fasta','a+')
    # fastafile1=open(paths+'LTR-Copia.fasta','a+')
    # fastafile2=open(paths+'DNA-En-Spm.fasta','a+')

    # refamilyfile0=open(referencePath+'LTR-Gypsy.txt','r')
    # refamilyfile1=open(referencePath+'LTR-Copia.txt','r')
    # refamilyfile2=open(referencePath+'DNA-En-Spm.txt','r')
    # refastafile0=open(referencePath+'LTR-Gypsy.fasta','r')
    # refastafile1=open(referencePath+'LTR-Copia.fasta','r')
    # refastafile2=open(referencePath+'DNA-En-Spm.fasta','r')

    os.system("cat "+referencePath+'LTR-Gypsy.txt'+">> "+paths+'LTR-Gypsy.txt')
    os.system("cat "+referencePath+'LTR-Copia.txt'+">> "+paths+'LTR-Copia.txt')
    os.system("cat "+referencePath+'DNA-En-Spm.txt'+">> "+paths+'DNA-En-Spm.txt')

    os.system("cat "+referencePath+'LTR-Gypsy.fasta'+">> "+paths+'LTR-Gypsy.fasta')
    os.system("cat "+referencePath+'LTR-Copia.fasta'+">> "+paths+'LTR-Copia.fasta')
    os.system("cat "+referencePath+'DNA-En-Spm.fasta'+">> "+paths+'DNA-En-Spm.fasta')





if __name__ == '__main__':
    # 获取存取path的txt文件
    p='/home/ubuntu/data/ltr_classification/classification1'
    paths ='/home/ubuntu/data/ltr_classification/classification1/'
    familys=["LTR/Gypsy","LTR/Copia","DNA/En-Spm"]

    listfile=open(paths+'list.txt','r')
    referencePath='/home/ubuntu/data/rubella/'

    familyfile0=open(paths+'LTR-Gypsy.txt','w')
    familyfile1=open(paths+'LTR-Copia.txt','w')
    familyfile2=open(paths+'DNA-En-Spm.txt','w')

    familyfile0.write("Sample"+'\t'+"Group")
    familyfile1.write("Sample"+'\t'+"Group")
    familyfile2.write("Sample"+'\t'+"Group")

  
    # getReFastaInfor(referencePath,familys)


    # for line in listfile:
    #     # 处理一个类群 
    #     # path:对应类群路径
    #     # paths：数据存储路径
    #     path=str(line)

    #     # 将类群x的对应ltr信息提取到paths/fasta
    #     getFastaInfor(path,paths,familys)


       
       
    #     print("success:"+str(line))
    # # 将参考基因组的fasta内容加入
    # writeReference(referencePath,paths)

    os.system("bash /home/ubuntu/Arabidopsis/Scripts/tree_construction.sh "+p+" LTR-Gypsy")
    # os.system("bash /home/ubuntu/Arabidopsis/Scripts/tree_construction.sh "+p+" LTR-Copia")
    # os.system("bash /home/ubuntu/Arabidopsis/Scripts/tree_construction.sh "+p+" DNA-En-Spm")

    # #修改名字
    # os.system("sudo mv test.txt new.txt")

```

## 生物信息学原理-homework5
+ 没有可视化，已完成
``` python
## 后端处理
def ATCGdeal(a):
    if(a=='A' or a=='a'):
        return 'T'
    if(a=='T' or a=='t'):
        return 'A'
    if(a=='C' or a=='c'):
        return 'G'
    if(a=='G' or a=='g'):
        return 'C'
    else:
        return 'X'

#
def get2line(sequence):

    # 可能需要先处理sequence
    if(sequence.startswith('>')):
        sequence1=str(sequence.split('\n',1)[1])
    else:
        sequence1=sequence

    sequences=list(sequence1)
    sequence2=""
    for s in sequences:
        m=ATCGdeal(s)
        sequence2=''.join([sequence2,m])

    sequence2=sequence2[::-1]
    sequence1=sequence1.upper()
    return sequence1,sequence2



def judgeAn(j):
    file=open('/home/ubuntu/Arabidopsis/Arabidopsis_sequence/test/An.txt','r')
    for line in file:
        a=line.split(' ')
        b=a[1].split('\n')
        if(j==str(a[0])):
            return str(b[0])
    else:
        return 'X'
def prointSequence(sequence1,sequence2):
    test1=list(sequence1+' ')
    test2=list(sequence2+' ')
    j1=""
    j2=""
    j3=""
    out1=""
    out2=""
    out3=""
    i=0
    i1=0
    i2=0
    for p,q in enumerate(test1):
        # read1
        if(p==0 or i==1):
            if(len(j1)==3):
                # out1=out1.join(judgeAn(j1))
                out1=''.join([out1,judgeAn(j1)])
                j1=""
            i=1
            j1=''.join([j1,str(q)])

        if(p==1 or i1==1):
            if(len(j2)==3):
                # out1=out1.join(judgeAn(j1))
                out2=''.join([out2,judgeAn(j2)])
                j2=""
            i1=1
            j2=''.join([j2,str(q)])

        if(p==2 or i2==1):
            if(len(j3)==3):
                # out1=out1.join(judgeAn(j1))
                out3=''.join([out3,judgeAn(j3)])
                j3=""
            i2=1
            j3=''.join([j3,str(q)])
        

    print(out1)
    print(out2)
    print(out3)

    j1=""
    j2=""
    j3=""
    out1=""
    out2=""
    out3=""
    i=0
    i1=0
    i2=0
    for p,q in enumerate(test2):
    # read1
        if(p==0 or i==1):
            if(len(j1)==3):
                # out1=out1.join(judgeAn(j1))
                out1=''.join([out1,judgeAn(j1)])
                j1=""
            i=1
            j1=''.join([j1,str(q)])

        if(p==1 or i1==1):
            if(len(j2)==3):
                # out1=out1.join(judgeAn(j1))
                out2=''.join([out2,judgeAn(j2)])
                j2=""
            i1=1
            j2=''.join([j2,str(q)])

        if(p==2 or i2==1):
            if(len(j3)==3):
                # out1=out1.join(judgeAn(j1))
                out3=''.join([out3,judgeAn(j3)])
                j3=""
            i2=1
            j3=''.join([j3,str(q)])
    print(out1)
    print(out2)
    print(out3)


if __name__ == '__main__':
    # sequence='MMtccgtagcccttaattt'
    sequence= input("please enter the data: ")
    sequence1,sequence2=get2line(sequence)
    prointSequence(sequence1,sequence2)   

```
# 4/5
## g4
+ 完成xml_modify代码
``` python
import re
import os
from TE_pipeline import randomSample
from TE_pipeline import sortSample
def infoget(group_number,outfile, ltr_family_getfile):
    infile=open(outfile,'r')
    # famliy=[]
    q=0
    outfile1=open(ltr_family_getfile,'w')
    # pattern0=r'[|](.*?)[|]'
    for line in infile:
        outfile1.write(str(group_number)+'\t')
        # 拆分每一行的数  split
        splitresult=line.split('\t')
        test0 = "".join(splitresult[0])
        outfile1.write(test0+'\t')
        # 提取ltr家族名称
        test1 = "".join(splitresult[1])
        test2=test1[::-1]
        test3=str(test2.split('|')[2])

        # famliy.append(re.findall(pattern0, test2)[0])
        # test3 = "".join(famliy[q])
        test4=test3[::-1]
        outfile1.write(test4+'\n')
        q=q+1
    infile.close()
    outfile1.close()  


if __name__ == '__main__':
    paths =os.walk(r'/home/ubuntu/data/Arabidopsis_sequence')
    for path, dir_lst, file_lst in paths:
        for dir_name in dir_lst:
            # 输入拟南芥的编号
            group_number=os.path.join(dir_name)
            # 初始的染色体number
            chr_number=1
            # 每次的路径
            filepath=os.path.join(path, dir_name)+"/"
            while chr_number<=5:

                fastafile1=filepath+"chr"+str(chr_number)+".ltr.fasta"
                ltr_identify=filepath+"chr"+str(chr_number)+".ltr_identify.xml"
                
                ltr_family_getfile=filepath+"chr"+str(chr_number)+".ltr.xml"
                outfile=filepath+"chr"+str(chr_number)+"-ltr.xml"

                random_xml_file=filepath+"chr"+str(chr_number)+".random-ltr.xml"
                random_file=filepath+"chr"+str(chr_number)+".random.ltr.xml"
                if(os.path.exists(ltr_family_getfile)):
                    os.remove(ltr_family_getfile)

                if(os.path.exists(random_file)):
                    os.remove(random_file)
                if(os.path.exists(fastafile1)==False):
                    
                    print("not-exsits:"+str(group_number)+"-"+str(chr_number))    
                    chr_number=chr_number+1  
                    continue       
                # blast
                os.system("blastn -query "+fastafile1+" -db /home/ubuntu/data/softwares/tedb -outfmt 6 > "+ltr_identify)
                os.system("awk '!a[$1]++{print}' "+ltr_identify+" > "+outfile)
                infoget(group_number,outfile, ltr_family_getfile)
                # os.remove(ltr_identify)
                os.remove(outfile)

                # sample
                if(os.path.exists(ltr_family_getfile)==True):
                    # step7 对每个xml文件进行随机抽样
                    randomSample(ltr_family_getfile,random_xml_file)
                    # step8 按转座子顺序重排文件
                    sortSample(random_xml_file,random_file)
                    os.remove(random_xml_file)
                    print("susscss:"+str(group_number)+"-"+str(chr_number))
                chr_number=chr_number+1
 ```


## 复习微生物分析
  + 文献阅读（https://link.springer.com/article/10.1007/s13238-020-00724-8
![avatar]( https://media.springernature.com/full/springer-static/image/art%3A10.1007%2Fs13238-020-00724-8/MediaObjects/13238_2020_724_Fig1_HTML.png?as=webp)

# 4/7
## 文献阅读
https://www.pnas.org/doi/10.1073/pnas.1921719117
+ TE和基因组大小呈正相关；没被甲基化的CpG/all CpG和基因组大小呈负相关
+ TE DNA的甲基化允许基因组扩增，也为基于TE的调控位点进行基因控制带来了新的机会


https://www.nature.com/articles/s41467-020-14995-6

https://mobilednajournal.biomedcentral.com/articles/10.1186/s13100-020-00223-x


# 4/8
## g4
+ 代码 ltr_classification
``` python
import os

import re
import random
import subprocess
def getLtrList(paths,path,familys):
    # 创建文件存储转座子的信息
    familyfile0=open(paths+str(familys[0])+'.txt','w')
    familyfile1=open(paths+str(familys[1])+'.txt','w')
    familyfile2=open(paths+str(familys[2])+'.txt','w')
    chr_number=1
    while chr_number<=5:
        xmlfile=open(path+'/chr'+str(chr_number)+'.ltr.xml','r')
        for line in xmlfile:
            test=line.split('\t')
            test0=str(test[2]).split('\n')
            if(str(test0[0])==str(familys[0])):
                familyfile0.write(str(test[1])+'\n')

            if(str(test0[0])==str(familys[1])):
                familyfile1.write(str(test[1])+'\n') 

            if(str(test0[0])==str(familys[2])):
                familyfile2.write(str(test[1])+'\n')

        chr_number+=1
    familyfile0.close()
    familyfile1.close()
    familyfile2.close()
    xmlfile.close()


def getLtrFastaInfo(paths,path):
    fastafile0=open(paths+'LTR-Gypsy.fasta','w')
    fastafile1=open(paths+'LTR-Copia.fasta','w')
    fastafile2=open(paths+'DNA-En-Spm.fasta','w')

    familyfile0=open(paths+str(familys[0])+'.txt','r')
    familyfile1=open(paths+str(familys[1])+'.txt','r')
    familyfile2=open(paths+str(familys[2])+'.txt','r')
    # familyfile0=open(familyfile0,'r')
    # familyfile1=open(familyfile1,'r')
    # familyfile2=open(familyfile2,'r')

    chr_number=1

    while chr_number<=5:
        fastafile=open(path+'/chr'+str(chr_number)+'.ltr.fasta','r')
        x=str(path).split('/')
        sequence_number=str(x[5])
        
        for line in fastafile:
            if line.startswith('>'):
                a=line.split('>')
                continue

            # for line0 in familyfile0:
            #     if(str(line0)==str(a[1])):
            #         #写入文件 
            #         fastafile0.write('>'+sequence_number+'-'+str(a[0])+str(line0))

            # for line1 in familyfile1:
            #     if(str(line1)==str(a[1])):
            #         fastafile1.write('>'+sequence_number+'-'+str(a[0])+str(line1))
            # for line2 in familyfile2:
            #     if(str(line2)==str(a[1])):
            #         fastafile2.write('>'+sequence_number+'-'+str(a[0])+str(line2))
        chr_number=chr_number+1
    fastafile.close()
    familyfile0.close()
    familyfile1.close()
    familyfile2.close()

def getFastaInfor(path,paths,familys):



        


    # 创建文件存储转座子的信息
    path=str(path.split('\n')[0])
    familyfile0=open(paths+str(familys[0])+'.txt','a')
    familyfile1=open(paths+str(familys[1])+'.txt','a')
    familyfile2=open(paths+str(familys[2])+'.txt','a')
    familyfile3=open(paths+str(familys[3])+'.txt','a')
    familyfile4=open(paths+str(familys[4])+'.txt','a')
    familyfile5=open(paths+str(familys[5])+'.txt','a')
    familyfile6=open(paths+str(familys[6])+'.txt','a')
    familyfile7=open(paths+str(familys[7])+'.txt','a')
    familyfile8=open(paths+str(familys[8])+'.txt','a')
    familyfile9=open(paths+str(familys[9])+'.txt','a')



    fastafile0=open(paths+str(familys[0])+'.fasta','a')
    fastafile1=open(paths+str(familys[1])+'.fasta','a')
    fastafile2=open(paths+str(familys[2])+'.fasta','a')   
    fastafile3=open(paths+str(familys[3])+'.fasta','a')   
    fastafile4=open(paths+str(familys[4])+'.fasta','a')   
    fastafile5=open(paths+str(familys[5])+'.fasta','a')   
    fastafile6=open(paths+str(familys[6])+'.fasta','a')   
    fastafile7=open(paths+str(familys[7])+'.fasta','a')   
    fastafile8=open(paths+str(familys[8])+'.fasta','a')   
    fastafile9=open(paths+str(familys[9])+'.fasta','a')   



    chr_number=1
    while chr_number<=5:
        ltr_number=''
        ltr_number0=''
        xmlfile=path+'/chr'+str(chr_number)+'.ltr.xml'
        fastafile=path+'/chr'+str(chr_number)+'.ltr.fasta'
        with open(xmlfile,'r') as f1, open(fastafile,'r') as f2:
            for l2 in f2:
                if(l2.startswith(">")):


                    if(ltr_number==ltr_number0):
                        l1 = f1.readline()
                        if(l1==''):
                            break
                        
                        ltr_number0=str(l1.split("\t")[1])
                        fa_number0=str(l1.split("\t")[2])
                        fa_number=str(fa_number0.split('\n')[0])
                        group_number=str(l1.split("\t")[0])

                    test=str(l2.split(">")[1])
                    ltr_number=str(test.split('\n')[0])
                        

                    if(ltr_number==ltr_number0):
                        if(fa_number==str(familys[0])):
                            fastafile0.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile0.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                            

                        if(fa_number==str(familys[1])):
                            fastafile1.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile1.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')

                        if(fa_number==str(familys[2])):
                            fastafile2.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile2.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                        if(fa_number==str(familys[3])):
                            fastafile3.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile3.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                        if(fa_number==str(familys[4])):
                            fastafile4.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile4.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                        if(fa_number==str(familys[5])):
                            fastafile5.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile5.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                        if(fa_number==str(familys[6])):
                            fastafile6.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile6.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')

                        if(fa_number==str(familys[7])):
                            fastafile7.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile7.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')

                        if(fa_number==str(familys[8])):
                            fastafile8.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile8.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')

                        if(fa_number==str(familys[9])):
                            fastafile9.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile9.write(group_number+'-'+ltr_number+'\t'+'A'+'\n')
                        continue
                    else:
                        continue

                if(ltr_number==ltr_number0):
                    if(fa_number==str(familys[0])):
                        fastafile0.write(l2)
                    if(fa_number==str(familys[1])):
                        fastafile1.write(l2)                    
                    if(fa_number==str(familys[2])):
                        fastafile2.write(l2)

                    if(fa_number==str(familys[3])):
                        fastafile3.write(l2)
                    if(fa_number==str(familys[4])):
                        fastafile4.write(l2)                    
                    if(fa_number==str(familys[5])):
                        fastafile5.write(l2)

                    if(fa_number==str(familys[6])):
                        fastafile6.write(l2)
                    if(fa_number==str(familys[7])):
                        fastafile7.write(l2)                    
                    if(fa_number==str(familys[8])):
                        fastafile8.write(l2)
                    if(fa_number==str(familys[9])):
                        fastafile9.write(l2)
        chr_number=chr_number+1

def getReFastaInfor(paths,familys):
    # 创建文件存储转座子的信息
    familyfile0=open(paths+str(familys[0])+'.txt','w')
    familyfile1=open(paths+str(familys[1])+'.txt','w')
    familyfile2=open(paths+str(familys[2])+'.txt','w')
    familyfile3=open(paths+str(familys[3])+'.txt','w')
    familyfile4=open(paths+str(familys[4])+'.txt','w')
    familyfile5=open(paths+str(familys[5])+'.txt','w')
    familyfile6=open(paths+str(familys[6])+'.txt','w')
    familyfile7=open(paths+str(familys[7])+'.txt','w')
    familyfile8=open(paths+str(familys[8])+'.txt','w')
    familyfile9=open(paths+str(familys[9])+'.txt','w')



    fastafile0=open(paths+str(familys[0])+'.fasta','w')
    fastafile1=open(paths+str(familys[1])+'.fasta','w')
    fastafile2=open(paths+str(familys[2])+'.fasta','w')   
    fastafile3=open(paths+str(familys[3])+'.fasta','w')   
    fastafile4=open(paths+str(familys[4])+'.fasta','w')   
    fastafile5=open(paths+str(familys[5])+'.fasta','w')   
    fastafile6=open(paths+str(familys[6])+'.fasta','w')   
    fastafile7=open(paths+str(familys[7])+'.fasta','w')   
    fastafile8=open(paths+str(familys[8])+'.fasta','w')   
    fastafile9=open(paths+str(familys[9])+'.fasta','w')   

    ltr_number=''
    ltr_number0=''

    xmlfile=paths+'rubella.ltr.xml'
    fastafile=paths+'.ltr.fasta'
    with open(xmlfile,'r') as f1, open(fastafile,'r') as f2:
        for l2 in f2:
                if(l2.startswith(">")):


                    if(ltr_number==ltr_number0):
                        l1 = f1.readline()
                        if(l1==''):
                            break
                        
                        ltr_number0=str(l1.split("\t")[1])
                        fa_number0=str(l1.split("\t")[2])
                        fa_number=str(fa_number0.split('\n')[0])
                        group_number=str(l1.split("\t")[0])

                    test=str(l2.split(">")[1])
                    ltr_number=str(test.split('\n')[0])
                        

                    if(ltr_number==ltr_number0):
                        if(fa_number==str(familys[0])):
                            fastafile0.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile0.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')
                            

                        if(fa_number==str(familys[1])):
                            fastafile1.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile1.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')

                        if(fa_number==str(familys[2])):
                            fastafile2.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile2.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')
                        if(fa_number==str(familys[3])):
                            fastafile3.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile3.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')
                        if(fa_number==str(familys[4])):
                            fastafile4.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile4.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')
                        if(fa_number==str(familys[5])):
                            fastafile5.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile5.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')
                        if(fa_number==str(familys[6])):
                            fastafile6.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile6.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')

                        if(fa_number==str(familys[7])):
                            fastafile7.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile7.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')

                        if(fa_number==str(familys[8])):
                            fastafile8.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile8.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')

                        if(fa_number==str(familys[9])):
                            fastafile9.write(">"+group_number+'-'+ltr_number+'\n')
                            familyfile9.write(group_number+'-'+ltr_number+'\t'+'B'+'\n')

                        continue
                    else:
                        continue

                if(ltr_number==ltr_number0):
                    if(fa_number==str(familys[0])):
                        fastafile0.write(l2)
                    if(fa_number==str(familys[1])):
                        fastafile1.write(l2)                    
                    if(fa_number==str(familys[2])):
                        fastafile2.write(l2)

                    if(fa_number==str(familys[3])):
                        fastafile3.write(l2)
                    if(fa_number==str(familys[4])):
                        fastafile4.write(l2)                    
                    if(fa_number==str(familys[5])):
                        fastafile5.write(l2)

                    if(fa_number==str(familys[6])):
                        fastafile6.write(l2)
                    if(fa_number==str(familys[7])):
                        fastafile7.write(l2)                    
                    if(fa_number==str(familys[8])):
                        fastafile8.write(l2)
                    if(fa_number==str(familys[9])):
                        fastafile9.write(l2)










def writeReference(referencePath,paths,familys):
    sampleFasta(referencePath,familys)
    for family in familys:

        
        os.system("cat "+referencePath+str(family)+'.txt'+">> "+paths+str(family)+'.txt')
        os.system("cat "+referencePath+'random.'+str(family)+'.fasta'+">> "+paths+'random.'+str(family)+'.fasta')


def sampleFasta(paths,familys):

    for family in familys:

        fastafile0=open(paths+str(family)+'.fasta','r')
        fastafile1=open(paths+'random.'+str(family)+'.fasta','w')
        txtfile=paths+str(family)+'.txt'

        out=subprocess.getoutput("wc -l %s" % txtfile)
        N=int(out.split()[0])
        i=0
        j=0
        if(N<=30):
            list=random.sample(range(1,N+1), N)
            list0=sorted(list)
        else:
            list=random.sample(range(1,N), 30)
            list0=sorted(list)
        for line in fastafile0:
            if(line=='\n'):
                break
            if(i==30):
                break
            if(line.startswith('>')):
                j=j+1
                if(int(list0[i])==j):
                    fastafile1.write(line)

            else:
                if(int(list0[i])==j):
                    fastafile1.write(line)
                    i=i+1
        fastafile0.close()
        fastafile1.close()
     
                






if __name__ == '__main__':
    # 获取存取path的txt文件
    p='/home/ubuntu/data/ltr_classification/admixed'
    paths ='/home/ubuntu/data/ltr_classification/admixed/'
    # familys=['ATHILA','ATREP18',"ATHILA2","ATHILA6A","ATDNA12T3_2","ATGP2","ATLANTYS1","ATHILA3","ATHILA4","ATLANTYS2","ATGP3","ATGP1","META1","HELITRONY1D","ATHILA6B","HELITRONY3","ATLANTYS3","ATCOPIA78","ATHILA4A","ARNOLD3"]
    familys=["ARNOLDY2","ATDNA12T3_2","ATGP1","ATGP10","ATGP7","ATHILA4A","ATHILA6B","ATHILA7","ATREP18","ATREP3"]
    listfile=open(paths+'list.txt','r')
    referencePath='/home/ubuntu/data/rubella/'

    # familyfile0=open(paths+str(familys[0])+'.txt','w')
    # familyfile1=open(paths+str(familys[1])+'.txt','w')
    # familyfile2=open(paths+str(familys[2])+'.txt','w')
    
    # familyfile0.write("Sample"+'\t'+"Group\n")
    # familyfile1.write("Sample"+'\t'+"Group\n")
    # familyfile2.write("Sample"+'\t'+"Group\n")
    # familyfile0.close()
    # familyfile1.close()
    # familyfile2.close()  


    # 处理参考基因组
    getReFastaInfor(referencePath,familys)

    # 循环创建文件
    for family in familys:

        folder =open(paths+family+'.fasta','w')
        folder.close()
        folder1=open(paths+family+'.txt','w')
        folder1.close()


    for line in listfile:
    #     # 处理一个类群 
    #     # path:对应类群路径
    #     # paths：数据存储路径
        path=str(line)

    #     # 将类群x的对应ltr信息提取到paths/fasta
        getFastaInfor(path,paths,familys)

  
        print("success:"+str(line))

    for family in familys:

        folder =open(paths+'random.'+family+'.fasta','w')
        folder.close()

    # 抽样 
    sampleFasta(paths,familys)

    # # 将参考基因组的fasta内容加入
    writeReference(referencePath,paths,familys)

    
    for family in familys:
        txtfile=paths+str(family)+'.txt'
        testfile=paths+str(family)+'x.txt'
        # 在txt文件第一行插入内容
        os.system("sed '1i Sample\tGroup' "+txtfile+' > '+testfile)
        os.remove(txtfile)
        os.system('mv '+testfile+' '+txtfile)
        # 建树
        os.system("bash /home/ubuntu/Arabidopsis/Scripts/tree_construction.sh "+p+" random."+str(family))
        #修改名字
        os.system("sudo mv "+paths+str(family)+".treefile "+paths+str(family)+".tree")
        
```

## 文献阅读
已完成


# 4/9
## g4
+ 进化树无法正常建树
  + 原因：fasta文件中有重复的转座子
    + list文件重复
  + 解决：重新跑list文件
## 夏令营整理
+ 上海科技大学
+ 北大生科院
+ 浙大生研院
+ 北京生命科学研究所
+ 个人陈述
  + 我是四川大学生命科学学院计算生物学专业的学生，主修方向为生物科学，辅修方向为软件工程。我在过去三年里认真钻研专业课基本知识及理论，系统地学习了统计学和生物信息学的知识，我的几门核心必修课均在 90 及 90 以上：植物生物学（97），结构生物学（93），数学建模与实验（93），随机过程（96），统计模型与模拟（100），概率统计（90）。我以 546 的分数通过了六级考试，且在后续两年中，我有意识地阅读英文文献以完成课程作业，目前英语水平已有了较大提升。
    经过大一学年的基础知识学习后，我希望能够实践生物信息学手段，于是我加入刘志斌老师课题组，并进行微生物高通量测序分析和植物基因组序列分析。2020年，我专注于土壤微生物高通量测序结果分析。通过对不同氮肥浓度、是否隔绝外围土壤和有无种植菠菜下的土壤微生物分析，我们发现氮肥通过影响非运动细菌的搭便车机制导致微生物群落的变化，从细胞运动性解释了氮肥浓度对植物生长的影响。在项目的实施过程中，我们遇到了根际细菌t检验的结果大幅度偏离正常值的情况，通过请教老师和师兄师姐，以及不断学习补充相关知识，并不断进行尝试，发现了数据异常点并重新清洗数据，最终完成了该项研究，并撰稿完成，已送审《》期刊，其中我位列第三作者。
    通过上一个课题，我学习了如何通过统计学方法对高通量测序的得到的OUT表，对α、β多样性、物种组成、组间差异检验和微生物网络分析有了更深层次的认识，生物信息学分析技能得到了较大的提升。但此时我仍不满足于此，希望能够挑战更高难度的课题。于是从2021年9月开始，我一直在进行基于Genome1001计划的拟南芥全基因组研究。我们利用种群遗传学在全球范围内重建拟南芥基因组的变化历史。我的主要工作是G4预测和转座子活性研究：利用Quadron对潜在的G4结构完成预测和筛选，并结合逆转录转座子的预测结果计算两者的重合度，并把ltr_finder预测的转座子结果按照拟南芥类别和转座子家族分类，均与拟南芥的参考基因组rubella构建进化树并完成可视化，分析不同转座子家族的转座子活性。
    我计划在本科四年级尽全力对目前手头的两个项目进行收尾和交接。在毕业离校到正式进入研究生阶段期间，对本科所涉猎的统计学和序列分析知识进行梳理总结，同时对基本的实验操作进行学习。在研究生阶段，我希望能够更全面、更深入地学习群体基因组的相关知识；如果有机会，我希望能参与实验和生物信息学相结合的项目。在充分了解前人研究成果的基础上，对有兴趣的、有研究价值的科研方向进行大胆实践，并挑战艰深难题。


# 4/10
## g4
``` r
# tree-beautify
library(ggplot2)
library(ggtree)
library(treeio)
library(tidytree)
library(cowplot)


tree <- read.tree("C:/Users/meansy/ATREP3.tree")
group_file <- read.table("C:/Users/meansy/ATREP3.txt",header = T,row.names = 1)
# 按类分组
groupInfo <- split(row.names(group_file), group_file$Group)
tree <- groupOTU(tree, groupInfo)
# 绘制进化树
#p1=ggtree(tree, layout="equal_angle", ladderize = FALSE, branch.length = "none",aes(color=group))+ theme(legend.position = "right")+ ggtitle("ATREP3")
p1=ggtree(tree, layout="equal_angle", ladderize = FALSE, branch.length = "none",aes(color=group))+ theme(legend.position = "right")+
    scale_color_manual(values = c("#72AD1E","#EBB003"))
p1 <- ggplot(heightweight, aes(x = ageYear, y = heightIn)) +geom_point()
p1=p1 +ggtitle("ATREP3") +theme(plot.title = element_text(hjust = 0.5,size=11)) #设置标题居中



tree <- read.tree("C:/Users/meansy/ATHILA7.tree")
group_file <- read.table("C:/Users/meansy/ATHILA7.txt",header = T,row.names = 1)
# 按类分组
groupInfo <- split(row.names(group_file), group_file$Group)
tree <- groupOTU(tree, groupInfo)
# 绘制进化树
#p2=ggtree(tree, layout="equal_angle", ladderize = FALSE, branch.length = "none",aes(color=group))+ theme(legend.position = "right")+ ggtitle("ATHILA7")
p2=ggtree(tree, layout="equal_angle", ladderize = FALSE, branch.length = "none",aes(color=group))+
    scale_color_manual(values = c("#72AD1E","#EBB003"))

p2 <- ggplot(heightweight, aes(x = ageYear, y = heightIn)) +geom_point()
p2=p2+ggtitle("ATHILA7") +theme(plot.title = element_text(hjust = 0.5,size=11)) #设置标题居中




tree <- read.tree("C:/Users/meansy/ATREP18.tree")
group_file <- read.table("C:/Users/meansy/ATREP18.txt",header = T,row.names = 1)
# 按类分组
groupInfo <- split(row.names(group_file), group_file$Group)
tree <- groupOTU(tree, groupInfo)
# 绘制进化树
#p3=ggtree(tree, layout="equal_angle", ladderize = FALSE, branch.length = "none",aes(color=group))+ theme(legend.position = "right")+ ggtitle("ATREP18")
p3=ggtree(tree, layout="equal_angle", ladderize = FALSE, branch.length = "none",aes(color=group))+ theme(legend.position = "right")+
    scale_color_manual(values = c("#72AD1E","#EBB003"))
p3 <- ggplot(heightweight, aes(x = ageYear, y = heightIn)) +geom_point()
p3=p3 +ggtitle("ATREP18") +theme(plot.title = element_text(hjust = 0.5,size=11)) #设置标题居中


#plot_grid(p1,p2,p3,ncol=4, labels=c("A","B","C"))
plot_grid(p1,p2,p3,ncol=3, rel_widths = c(0.4, 0.4,0.4),align='hv')
                                            
```

## 夏令营
**个人陈述**
尊敬的评审老师：
您好！
感谢您拨冗审核我的个人陈述。我叫夏琦妮，目前就读于四川大学生命科学学院计算生物学交叉试验班，主修方向为生物科学，辅修方向为软件工程。为了使老师更好地了解我的情况，我将从教育背景、科研经历和未来计划三个方面进行陈述：
1.教育背景
我在过去三年里认真钻研专业课基本知识及理论，系统地学习了统计学和生物信息学的知识，我的几门核心必修课均在 90 及 90 以上：植物生物学（97），结构生物学（93），数学建模与实验（93），随机过程（96），统计模型与模拟（100），概率统计（90）。本人具有一定的英语功底，分别以553分和546分的成绩通过了英语四六级考试，在平时的学习中，我有意识地阅读英文文献以完成课程作业，目前英语水平已有了较大提升。
同时，我主动参与学科竞赛以丰富自身知识体系，并获得四川省大学生生物与环境科技创新大赛二等奖；主持省级大学生创新项目“氮肥浓度对土壤及根际微生物的重建研究”，并作为核心成员参与省级大学生创业项目“富硒酵母的评价及生物硒形态的分析”；本人还先后参与两届“互联网＋”大学生创新创业大赛，获校级三等奖。
为全面提升自身素质，我担任了校团委调查研究部副部长，多次组织参与四川大学学生思想动态调研并运营青春川大公众号，同时负责“四川大学青年研究课题”的开展，最终获得了“全国优秀高校校园媒体成员”称号。
2.科研经历
经过大一学年的基础知识学习后，我对植物生物学充满兴趣，并希望亲自通过生物信息学手段以揭示与植物相关的生物学现象，于是我加入了刘志斌老师课题组，并对与植物根际互作的土壤微生物进行高通量测序分析：借助基迪奥生物信息云平台对不同氮肥浓度、是否隔绝外围土壤（滤膜处理）和有无种植菠菜下的土壤、根际微生物进行α、β多样性、物种组成、微生物群落功能和微生物网络分析，我们发现：不同氮浓度下，无滤膜组根际微生物α、β多样性水平显著，细胞运动表现出极大变化，滤膜的加入减弱了不同氮浓度下细胞运动功能丰度的变化。以上说明氮肥通过影响非运动细菌的搭便车机制引起微生物群落的变化，从细胞运动学的角度解释了氮肥浓度对植物生长的影响。在项目的实施过程中，我们遇到了根际细菌t检验的结果大幅度偏离正常值的情况，通过请教老师和师兄师姐，以及不断学习补充相关知识，并不断进行尝试，发现了数据异常点并重新清洗数据，最终完成了该项研究，并撰稿完成，现处于“BIOLOGY AND FERTILITY OF SOILS”在投阶段，其中我位列第三作者。
通过上一个课题，我对统计学在生物领域的应用有了更深层次的认识，生物信息学分析技能得到了较大的提升。但我仍不满足于此，希望能够挑战更高难度的课题。于是从2021年9月开始，我和我的团队一直在进行基于Genome1001计划的拟南芥全基因组研究。我们期望利用种群遗传学在全球范围内重建拟南芥基因组的变化历史,探究不同地理环境下拟南芥基因组中G4、TE和甲基化的关系，并结合因果推断转座事件的先后关系，分析物种内部进化史。目前我已经完成的工作包括G4预测和转座子活性研究：利用Quadron对潜在的G4结构完成预测和筛选，并结合逆转录转座子的预测结果计算两者的重合度；将ltr_finder预测的转座子结果按拟南芥类别和转座子家族分类，均与拟南芥的参考基因组rubella构建系统发育树并完成可视化，从而分析不同转座子家族的转座子活性。
3.未来计划
经过两年的生信分析项目实践，我在该方面的能力得到了显著的提升。同时，在与负责湿实验同学的交流过程中，我慢慢对揭示、验证生物现象的湿实验感兴趣，计划在本科四年级对目前手头的两个项目进行收尾和交接，对本科所涉猎的统计学和序列分析知识进行梳理总结，同时学习湿实验的基本实验操作。
在研究生阶段，我希望能够更全面、更深入地学习群体基因组的相关知识；如果有机会，我希望能参与实验和生物信息学相结合的项目。在研究生第一年时认真学习，充分运用xx大学多学科、国际化、立足前沿的学科资源，积极收听感兴趣的前沿讲座，进一步提升自己的科研品味、开拓自身的科研视野，同时提升自己对国际上的前沿问题、国家“卡脖子”技术问题的认识。从研究生第二年起，我将与科研团队团结协作，按照实验室的战略方向以及制定的科研计划进行研习，在学习以及科研的过程在中虚心向实验室师兄师姐以及导师请教，在充分了解前人研究成果的基础上，对有兴趣的、有研究价值的科研方向进行大胆实践，并挑战艰深难题。
享誉盛名的xx大学在生物信息学领域处于国内领先水平，所以我非常希望在贵校xx院完成我的研究所课程，我会尽自己最大的努力投身于生物领域，成为贵学院研究团队里的一名优秀研究生！
我的个人陈述至此完毕，再次感谢评审老师，祝身体健康，事业顺利！
夏琦妮
2022年4月10日

## 作业
+ 中华文化
  + 【中期作业】
•中华文化 艺术篇中期作业：
通过2022春晚上的舞蹈“只此青绿”的鉴赏与了解，并结合中华文化艺术篇课堂教学内容，请你谈谈中国传统艺术精神及美学品格的魅力。
•内容要求：
1、题目自取。
2、围绕一至二个观点展开即可 。
3、务必联系具体艺术作品或案例谈论。
4、字数一千字左右。
•提交要求：
1、纸质版（写清学院＋学号＋姓名）
2、仅在第十周（4.27）课间时间收取，请勿提前或延后提交，也请勿在老师授课时交作业。
    + 创作动机 《千里江山图
最开始，当‘青绿’望月，垂目转身，把观众带入到《千里江山图》的画卷中，让这幅画活起来；最后，当一座座小山变幻成险峻陡峭的山峰，就如同‘会当凌绝顶，一览众山小’，透过画家王希孟的双眼，眺望祖国大好河山，仿佛在问观众：“你们看到古人笔下中华的大好河山了吗？看到古人对后人的馈赠了吗？我想，那一刻，每个人都能感受到，‘这很中国’。”
《千里江山图》中的矿物质颜料（石青、石绿）历经千年而未衰，即使在幽暗的环境中依然泛着宝石的光芒，主创团队正是从这一细节处提炼出“青绿”的意象。
既是历经岁月不褪色的璀璨耀眼，也是漫长时光沉寂后的隐忍和坚定。
青绿代表着中华优秀传统文化，没有小我，只有大我。


# 4/11
## bioinformatics
+ Burrows Wheeler Transform
  + https://zh.wikipedia.org/wiki/Burrows-Wheeler%E5%8F%98%E6%8D%A2

## 中华文化作业


## 日常
+ KA/KS 
  + https://zh.wikipedia.org/wiki/Ka/Ks


# 4/12
## ltr
+ 转座子活性
  + 进化树
  + KAKS
+ 转座子溯源
  + 距离计算
  + 年龄
    + https://genomebiology.biomedcentral.com/articles/10.1186/gb-2004-5-10-r79#Sec11

# 4/13
## bioinformatics
+ homework
    + BWT算法实现
    + 并比对猪瘟病毒序列



# 4/15
## 邮件
## g4
+ 根据距离矩阵绘制进化树
  + 绘制：http://www.biotrainee.com/thread-2394-1-1.html
  + 美化： https://abego.cn/2019/07/10/the-r-packages-ggtree-make-the-tree-more-possible/

## 个人陈述
个人陈述
尊敬的评审老师：
您好！
感谢您拨冗审核我的个人陈述。我叫夏琦妮，目前就读于四川大学生命科学学院计算生物学交叉试验班，主修方向为生物科学，辅修方向为软件工程。为了使老师更好地了解我的情况，我将从教育背景、科研经历和未来计划三个方面进行陈述：
1.教育背景
我在过去三年里认真钻研专业课基本知识及理论，系统地学习了统计学和生物信息学的知识，我的几门核心必修课均在 90 及 90 以上：植物生物学（97），结构生物学（93），数学建模与实验（93），随机过程（96），统计模型与模拟（100），概率统计（90）。本人具有一定的英语功底，分别以553分和546分的成绩通过了英语四六级考试，且在平时的学习中，我有意识地阅读英文文献以完成课程作业，目前英语水平已有了较大提升。
同时，我主动参与学科竞赛以丰富自身知识体系，并获得四川省大学生生物与环境科技创新大赛二等奖；主持省级大学生创新项目“氮肥浓度对土壤及根际微生物的重建研究”，并作为核心成员参与省级大学生创业项目“富硒酵母的评价及生物硒形态的分析”。
为全面提升自身素质，我担任校团委调查研究部副部长，多次组织参与四川大学学生思想动态调研并运营青春川大公众号，同时负责“四川大学青年研究课题”的开展，期间受到了红色思想感染，树立了马克思主义思想观点。
2.科研经历
经过大一学年的基础知识学习后，我对植物生物学充满兴趣，并希望亲自通过生物信息学手段以揭示与植物相关的生物学现象，于是我加入了刘志斌老师课题组，并对与菠菜种植下土壤微生物进行高通量测序结果的分析，我主要负责通过Omicsmart平台对不同氮肥浓度、是否隔绝外围土壤（滤膜处理）的土壤、根际微生物进行α、β多样性、物种组成、群落功能分析，以期从细胞运动学的角度阐述氮肥浓度对植物生长的影响机制。在这一过程中，我发现：不同氮浓度下，无滤膜组根际微生物多样性水平显著，而滤膜的加入减弱了细菌运动功能丰度的变化。结合团队成员的工作，最终发现氮肥通过影响非运动细菌的搭便车机制引起微生物群落的变化。在项目的实施过程中，我遇到了根际细菌t检验的结果大幅度偏离正常值的情况，通过不断学习补充相关知识，发现了数据异常点并重新清洗数据，最终完成了该项研究，并撰稿完成，现处于“BIOLOGY AND FERTILITY OF SOILS”在投阶段，其中我位列第三作者。
通过上一个课题，我对统计学在生物领域的应用有了更深层次的认识，生物信息学分析技能得到了较大的提升。但我仍不满足于此，希望能够挑战更高难度的课题。于是从2021年9月起，我参与了探究不同地理环境下拟南芥基因组中G4、转座子和甲基化关系的课题。我和师兄预测和筛选了潜在的G4，并完成转座子的预测，通过数学计算的手段发现：不同地区拟南芥的转座子、G4和拟南芥基因组大小均呈正相关；但含有G4的转座子数量和基因组大小没有相关性，我计划继续探究基因组中转座子和甲基化、G4的关系，以期深度剖析三者的关系，解释G4、甲基化对转座子的影响机制。从2022年1月起，我参与设计了在上一课题的基础上，通过因果推断转座事件的先后关系分析拟南芥物种内部进化史的新课题。我以转座子家族和拟南芥所处地区作为分类标准，同时考虑参考基因组，抽样构建转座子的系统发育树，从而分析不同转座子家族间的距离，以期结合年龄计算模型探究转座子的活性，同时作为参考完成转座子的溯源工作。
3.未来计划
经过两年的生信分析项目实践，我在该方面的能力得到了显著的提升。同时，在与负责湿实验同学的交流过程中，我慢慢对揭示、验证生物现象的湿实验感兴趣。计划在本科四年级对目前手头的两个项目进行收尾和交接，对本科所涉猎的统计学和序列分析知识进行梳理总结，同时学习湿实验的基本实验操作。
在研究生阶段，我希望能够更全面、更深入地学习群体基因组的相关知识；如果有机会，我希望能参与实验和生物信息学相结合的项目。在研究生第一年时认真学习，充分运用xx大学多学科、国际化的学科资源，开拓自身的科研视野，同时提升自己对国际上的前沿问题、国家“卡脖子”技术问题的认识。从研究生第二年起，我将与科研团队团结协作，按照实验室的战略方向和科研计划进行研习，在科研的过程在中虚心向实验室师兄师姐以及导师请教，在充分了解前人研究成果的基础上，对有兴趣的、有研究价值的科研方向进行大胆实践，并挑战艰深难题。
享誉盛名的xx大学在生物信息学领域处于国内领先水平，所以我非常希望在贵校xx院完成我的研究所课程，我会尽自己最大的努力投身于生物领域，成为贵学院研究团队里的一名优秀研究生！
我的个人陈述至此完毕，再次感谢评审老师，祝身体健康，事业顺利！

# 4/17
## bioinformatics
```python

import numpy as np
def BWTsequence(sequence0):
    sequence0='$'+sequence0
    # 创建存储矩阵
    i=0
    record0 = np.zeros((len(sequence0), len(sequence0)), dtype=str)
    list0=list(sequence0)
    while i<len(sequence0):

        t1=sequence0[-i:]
        t2=sequence0[:-i]
        t3=t1+t2
        list0=list(t3)
        z=0
        while z<len(sequence0):
            record0[i,z]=list0[z]
            z=z+1
        i=i+1
    # print(record0)
    return record0
    # print(record0)

def Matrix(record0):
    # 输入行和列
    x=record0.shape[0]
    y=record0.shape[1]

    i=1

    while i<x:
        j = i+1
        while j<x:
            m=0
            while m<x:
                if(record0[i,m]<record0[j,m]):
                    break
                if(record0[i,m]==record0[j,m]):
                    m=m+1
                    continue
                if(record0[i,m]>record0[j,m]):
                    tmp = np.copy(record0[i])
                    record0[i] = record0[j]
                    record0[j] = tmp
                    break

            j=j+1

        i=i+1
    # print(record0)
    p=0
    listF = np.zeros((0,))
    listL = np.zeros((0,))
    while p<x:
        listF = np.append(listF, record0[p,0])
        listL = np.append(listL, record0[p, x-1])
        p=p+1

    # print(listL)
    # print(listF)
    m=1
    n=1
    p=1
    q=1
    rankL = []
    for cc in range(0, len(listL)):
        if(str(listL[cc])=='a' or str(listL[cc])=='A'):
            rankL.append(m)
            m=m+1
        if(str(listL[cc])=='t' or str(listL[cc])=='T'):
            rankL.append(n)
            n=n+1
        if(str(listL[cc])=='c' or str(listL[cc])=='C'):
            rankL.append(p)
            p=p+1
        if(str(listL[cc])=='g' or str(listL[cc])=='G'):
            rankL.append(q)
            q=q+1
        if(str(listL[cc])=='$'):
            rankL.append(0)

    m = 1
    n = 1
    p = 1
    q = 1
    rankF = []
    for tt in range(0, len(listF)):
        if(str(listF[tt])=='a' or str(listF[tt])=='A'):
            rankF.append(m)
            m=m+1
        if(str(listF[tt])=='t' or str(listF[tt])=='T'):
            rankF.append(n)
            n=n+1
        if(str(listF[tt])=='c' or str(listF[tt])=='C'):
            rankF.append(p)
            p=p+1
        if(str(listF[tt])=='g' or str(listF[tt])=='G'):
            rankF.append(q)
            q=q+1
        if(str(listF[tt])=='$'):
            rankF.append(0)
    # print(rankL)
    # print(rankF)


    return listL,rankL,listF,rankF


def ranks(sequence):
    m = 1
    n = 1
    p = 1
    q = 1
    rankF = []
    listF=list(sequence)
    for tt in range(0, len(listF)):
        if(str(listF[tt])=='a' or str(listF[tt])=='A'):
            rankF.append(m)
            m=m+1
        if(str(listF[tt])=='t' or str(listF[tt])=='T'):
            rankF.append(n)
            n=n+1
        if(str(listF[tt])=='c' or str(listF[tt])=='C'):
            rankF.append(p)
            p=p+1
        if(str(listF[tt])=='g' or str(listF[tt])=='G'):
            rankF.append(q)
            q=q+1
        if(str(listF[tt])=='$'):
            rankF.append(0)
    return rankF




def BWT2deserve(listL,rankL,listF,rankF,sequence,re_sequence):
    #倒序
    sequence1=sequence[::-1]
    sequencex=list(sequence1)
    rerank = ranks(re_sequence)
    xx=0
    ww=0
    for m in range(len(listF)):
        if (str(listF[m]) == str(sequencex[0])):
            outsequence = ['-'] * len(re_sequence)
            test=[]
            out=''
            number=1
            qq=1
            gg=m
            rank=0

            test.append(sequencex[0])

            while len(re_sequence) > qq:

                if (len(sequencex)>qq and listL[gg] != sequencex[qq]):
                    test = []
                    break

                else:
                    for l1 in range(len(listF)):
                        if (rankL[gg] == rankF[l1] and listL[gg] == listF[l1]):
                            #print(rankL[gg])
                            gg = l1
                            test.append(listF[l1])
                            qq = qq + 1
                            break

                if (listF[l1] == '$'):
                    if(xx==0):
                        # print("reference_sequence:")
                         # print(re_sequence)
                        print("alignemnt_results:")
                        xx=xx+1
                    # 截取
                    j=len(test)-2
                    i=j-len(sequencex)+1

                    while i<=j:
                        ww=1
                        outsequence[i]=re_sequence[i]
                        i=i+1
                    out=''.join(outsequence)
                    print(out)
                    break
    if(ww==0):
        # print("reference_sequence:")
        # print(re_sequence)
        print("sorry,cannot find alignment sequence!")

# encode
def rle_compress(s):

    result = ''
    last = s[0]
    count = 1
    for _ in s[1:]:
        # 采用一个字符表示重复次数，所以count最大是9
        if last == _ and count < 9:
            count += 1
        else:
            result += str(count) + last
            last = _
            count = 1
    result += str(count) + last
    return result


def rle_decompress(s):
    result = ''
    for _ in range(len(s)):
        if _ % 2 == 0:
            result += int(s[_]) * s[_ + 1]
    return result

def getListF(listL):
    listF = np.zeros((0,))
    c = 0
    for m in range(len(listL)):


        if(listL[m]=='$' or c==1):
            c=1

            np.append(listF, listL[m])

    for j in range(len(listL)):
        if(listL[j]!='$'):
            np.append(listF, listL[j])
        else:
            break


    return listF




def BWTback( listL, rankL, listF, rankF):
    g=0
    test=[]
    tt=1
    # test.append(listL[0])
    while tt<=len(listL)-1:
        for i in range(len(listF)):
            if(listF[i]==listL[g] and rankL[g]==rankF[i]):
                tt=tt+1
                test.append(listL[g])
                g=i
        if(listL[g]=='$'):
            break
        # test.append(listL[g])
    test=test[::-1]
    print('reference_sequence(restore sequence):\n'+''.join(test))



def fasta2str(file1):
    infile=open(file1,'r')
    re_sequence=''
    for line in infile:
        if(line.startswith('>')):
            continue
        else:
            xx=line.split('\n')
            re_sequence=re_sequence+str(xx[0])
    re_sequence=re_sequence.lower()
    return re_sequence



if __name__ == '__main__':
    re_sequence = input("Enter a reference sequence(if want to use default sequence,input'0'):")
    if(re_sequence=='0'):
        re_sequence=fasta2str('D:\\新建文件夹\\大三下\\test.fasta')
    sequence = input("Enter a comparison sequence: ")
    sequence=sequence.lower()
    # re_sequence='actgagctttagcgtagctttaggagagcttcctagctacgtatcgagcgggcatctatc'
    # 矩阵 record0
    record0=BWTsequence(re_sequence)

    listL, rankL, listF, rankF=Matrix(record0)

    # # 游程编码压缩listL
    # listx=rle_compress(''.join(listL))
    # # 解压缩
    # codeL=rle_decompress(listx)
    # listL=list(codeL)
    # 由listL 得到lisF
    listx=getListF(listL)
    #还原序列
    sequenceo=BWTback( listL, rankL, listF, rankF)
    BWT2deserve(listL,rankL,listF,rankF,sequence,re_sequence)

```



# 4/18
## bioinformatics
+ 进化生物学
  + 宏观到微观
    + 宏观：化石记录 形态比较等
    + 微观：分析进化理论

+ 分子进化基础
  + 中性漂移
    + 突变随机  自然选择/进化压力有向
  + 分子钟理论（由中性漂移直接推出
    + 可以直接基于分子水平研究进化
+ 进化树（一般是二叉树
  + 树的类型
    + 怎么构建有根树的根？
      + 设置外围群     
  + **如何构建进化树？**
    + 选择:同源分子（相似性可以达到一定程度
    + 多序列比对，**截取最好区域作为代表**
    + 差异性计算
    + 基于差异度建树
      + 距离法（大量序列） 量化距离 构建距离矩阵
      + 最大简约法（近缘序列
      + 最大似然法（远缘序列
    + 可信度检验
+ 建树方法
  + 距离法
    + 贪婪算法 局部最优 但无法使得全局最优
  + 最大简约法
    + 遍历所有树
    + 假设**进化是向最节省的方向进化的**
      + 假设不一定成立 因此应该多数用于序列相似度非常高的序列
    + 求和 碱基突变总数 取min
  + 极大似然法
    + 和最大简约法不同的地方:是根据碱基替换概率模型（eg：氨基酸的pam矩阵），概率相乘，遍历所有树 max
    + 为什么是max？我们希望这棵树发生的概率最大
    + 计算量大，可以分成子树，这种方法其实也是一种贪婪算法
  + 贝叶斯法
+ 进化树的可信度检验
  + bootstrap
    +  序列相似性过高，反而bootstrap会很低
    +  因此在相似度很高的时候不用计算
